Ovsdb schema upgrade

mkalcok opened the pull request:
Please review each commit individually.

This change set lets MicroOVN to take control of the OVN OVSDB cluster schema upgrades.

It also includes few refactoring changes, that are in my opinion required.  The `microovn/ovn` package got bit bloated over time and it became hard to use its functionality from other packages. Importing `microovn/ovn` would often result in circular dependencies, so I attempted to extract, and group, functions that could stand on their own into standalone packages.

I consider the implementation good for review. Functional tests and documentation are in progress.

fnordahl commented on microovn/ovn/commands.go:
```diff
@@ -16,12 +16,14 @@ const defaultDBConnectWait = 30 //Default time to wait for connection to ovsdb
+	ShortName    string // Shorthand name for the database (e.g. "nb")
+	FriendlyName string // Human friendly name of the database ideal for logging purposes (e.g. "Northbound")
+	Schema       string // Path to a schema file for the database
+	IsCentral    bool   // Whether the database is used by OVN central services
```
Adding a bool for this purpose feels redundant since we already have a notion of OvsdbType.

Can we record the OvsdbType instead and perhaps call a helper function which knows which of the types are used for central DBs?

mkalcok replied:
I guess that this is a matter of perspective. If we replaced `IsCentral` with something like `DbType`, that would hold `OvsdbType`, then we would still have same number of members in the `OvsdbSpec`, but now we'd also need a helper method that would decide whether an instance of `OvsdbSpec` is central or not.

In a sense, we already have a helper like this. It's the constructor for the `OvsdbSpec`. You pass `OvsdbType` to `NewOvsdbSpec` and it fills in the `IsCentral` member with true/false.

fnordahl replied:
Yea, I guess it is a matter of preference of using methods that understand the data structure rather than accessing the data directly from foreign functions/method.

Right now it does not make much difference so I'm fine either way, but if we somewhere down the road need to make other decisions on the back of the different types we might find it makes more sense to store the type.

fnordahl commented on microovn/ovn/commands.go:
```diff
@@ -16,12 +16,14 @@ const defaultDBConnectWait = 30 //Default time to wait for connection to ovsdb
-	Name   string
+// OvsdbSpec is a helper structure that encapsulates properties of an OVN/OVS database.
+type OvsdbSpec struct {
+	SocketPath   string // Path to a socket that the database has open
```
While I see that the code is currently hard coded for Unix sockets, it feels limiting knowing the underlying tooling accepts any type of target.

OTOH, if we will only ever use this for sockets i guess its fine and we can always change it later.

mkalcok replied:
I see what you mean and it's a good point. The reason I wanted to change it from the previous name `Target` was that some commands, like `ovn-appctl`, take "target" argument which points to `.ctl` file as opposed to `.sock`. I wanted to avoid confusions there.

I'll think of some better name, and I'm open to suggestions. Perhaps a `LocalSocketPath`?

If we have ever a need for using tcp/ssl "targets", we can extend this struct later.

mkalcok replied:
Come to think about this, the `LocalSocketPath` wouldn't be the best option. If we ever need to use network targets for the database connection, we could probably add new `OvsdbType`, something like `OvsdbTypeNBCluster` and the constructor for `OvsdbSpec` would populate this field with correct tcp/ssl URLs. So perhaps this field could be called `SocketURL` and it would work for `unix:`, `tcp:` and `ssl:`

fnordahl commented on microovn/ovn/cli/commands.go:
```diff
@@ -1,4 +1,4 @@
-package ovn
+package cli
```
The name `cli` is a bit misleading, is this package implementing a CLI?

If the goal is to shorten the `commands` name, would `cmd` work as well?

mkalcok replied:
Thank you, yes. I had my doubts about the `cli` as well. The `cmd` alternative (with `ovn_cmd` alias usage) is much better.

fnordahl commented on microovn/ovn/bootstrap.go:
```diff
@@ -131,7 +132,7 @@ func Bootstrap(s *state.State, initConfig map[string]string) error {
 	}
 
-	_, err = VSCtl(
+	_, err = cli.VSCtl(
```
The imported name is a bit ambiguous, should we perhaps use the golang alias functionality and import the package with a prefix so that it could read `ovn_cmd.VSCtl()`

fnordahl reviewed (commented):
Thanks for your work on this, Martin! Have unfortunately not been able to help test, but the approach looks good to me.

Do we need some additional CLI knobs to allow the human operator to inform themselves about the progress? For example showing the expected schema version in cluster status might help them spot whether a snap upgrade is missing on one of the nodes?

mkalcok commented:
> Do we need some additional CLI knobs to allow the human operator to inform themselves about the progress? For example showing the expected schema version in cluster status might help them spot whether a snap upgrade is missing on one of the nodes?

I like this idea. It would probably took another API endpoint. But we could add information about the schema upgrade status. I'll think about way to present this to the user in `microovn status` output, because just showing expected DB schema version might not cut it. It might be confusing to display information about expected NB schema version on a node that doesn't run "central" service.

fnordahl commented on microovn/api/types/ovsdb.go:
```diff
@@ -0,0 +1,118 @@
+		}
+		msg += fmt.Sprintf("\t%s: ", nodeName)
+		if node.Error {
+			msg += "N/A\n"
```
This message is printed in multiple situations, and it does not appear to be particularly helpful for the operator.

1) For example when only one node of the cluster has been upgraded, the non-upgraded nodes will be shown as `N/A` regardless of them having central components that actually do need to be upgraded or not, example:
```
$ sudo microovn status
MicroOVN deployment summary:
- micro-1 (10.5.0.2)
  Services: central, chassis, switch

- micro-2 (10.5.0.3)
  Services: central, chassis, switch

- micro-3 (10.5.0.4)
  Services: central, chassis, switch

- micro-4 (10.5.0.5)
  Services: chassis, switch

OVN Database summary:
OVN Southbound: Upgrade or attention required!
Currently running schema: 20.21.0
Cluster report (expected schema versions):
    micro-1: 20.33.0
    micro-3: N/A
    micro-2: N/A
    micro-4: N/A

OVN Northbound: Upgrade or attention required!
Currently running schema: 6.1.0
Cluster report (expected schema versions):
    micro-1: 7.3.0
    micro-2: N/A
    micro-3: N/A
    micro-4: N/A
```

2)  When all central nodes are upgraded and the chassis nodes are not upgraded, the schema conversion will (rightfully) be held back, but the status does not reveal any information to the end user that this is the reason for the upgrade not happening.

```
$ sudo microovn status
MicroOVN deployment summary:
- micro-1 (10.5.0.2)
  Services: central, chassis, switch

- micro-2 (10.5.0.3)
  Services: central, chassis, switch

- micro-3 (10.5.0.4)
  Services: central, chassis, switch

- micro-4 (10.5.0.5)
  Services: chassis, switch

OVN Database summary:
OVN Southbound: Upgrade or attention required!
Currently running schema: 20.21.0
Cluster report (expected schema versions):
    micro-1: 20.33.0
    micro-4: N/A
    micro-3: 20.33.0
    micro-2: 20.33.0

OVN Northbound: Upgrade or attention required!
Currently running schema: 6.1.0
Cluster report (expected schema versions):
    micro-1: 7.3.0
    micro-4: N/A
    micro-3: 7.3.0
    micro-2: 7.3.0
```

So for the chassis node could we perhaps print a hint about snap upgrade being required? Another possiblitiy is to use the word "supported" instead of "expected" for schema version, that way it would not be confusing when we print schema version for the "client" nodes?

Mock-up:
```
Cluster report (supported schema versions):
    micro-1: 7.3.0
    micro-4: Unknown - snap upgrade required
    micro-3: 7.3.0
    micro-2: 7.3.0

Cluster report (supported schema versions):
    micro-1: 7.3.0
    micro-4: 6.1.0 - snap upgrade required
    micro-3: 7.3.0
    micro-2: 7.3.0
```

mkalcok replied:
Thank you for the review and testing @fnordahl. For starters I want to clarify that every node, regardless of whether it runs central services or not, is expected to report "expected version" for the schema. However, we are facing kind of a one-off situation here where non-upgraded nodes do not have the API required for this feature to work. Therefore any non-upgraded nodes report `N/A`, in the future these nodes would report older schema version.
We can probably distinguish `404` response code and print more informative message in this scenario. That would leave the `N/A` value solely for unexpected errors during the expected schema version fetch.

So in the current scenario, where some hosts in the cluster do not support this feature you'd see something like:

OVN Database summary:
OVN Southbound: Upgrade or attention required!
Currently running schema: 20.21.0
Cluster report (expected schema versions):
    micro-1: 20.33.0
    micro-4: API not supported - MicroOVN requires upgrade
    micro-3: 20.33.0
    micro-2: 20.33.0

fnordahl reviewed (commented):
I took this for a spin, and apart from the comment in-line, the ergonomics of this when knowing how it works is good, great job on that! If we resolve the remaining issues this will be a fantastic UX imho.

Note that I did not do any thorough validation of data plane interruptions so I trust that to work as expected.

mkalcok commented:
I updated this PR based on the discussions we had above @fnordahl. Key points:

Distinguishing between unreachable or otherwise faulty cluster members and cluster members that lack API to report expected schema version:
```
OVN Database summary:
OVN Southbound: Upgrade or attention required!
Currently running schema: 20.21.0
Cluster report (expected schema versions):
	movn1: 20.33.0
	movn4: Error. Failed to contact member
	movn3: Missing API. MicroOVN needs upgrade
	movn2: Missing API. MicroOVN needs upgrade

OVN Northbound: Upgrade or attention required!
Currently running schema: 6.1.0
Cluster report (expected schema versions):
	movn1: 7.3.0
	movn4: Error. Failed to contact member
	movn2: Missing API. MicroOVN needs upgrade
	movn3: Missing API. MicroOVN needs upgrade
```

Added current schema version to "OK" output:
```
OVN Database summary:
OVN Southbound: OK (20.33.0)
OVN Northbound: OK (7.3.0)
```

Added functional tests to verify that schema upgrade gets triggered only after every host in the cluster is upgraded.

mkalcok commented:
Bit of a side-note: I talked to you offline @fnordahl about adding the DB schema upgrade test into the `upgrade.bats` and reusing containers, but I went back on that idea. It didn't really saved much time as the time-consuming part is installing/bootstrapping MicroOVN, not creating containers. Other drawbacks:
* harder parallelization in CI
* Messy setup/teardown that needs to clean up container after test

It wasn't a good idea and we should stick with single-use containers in the tests.

mkalcok commented:
Thank you @fnordahl. Yes I had it cherry-picked in this branch and forgot about it. I'll drop it and re-submit

mkalcok commented:
Two changes since last push:

1. Dropped the accidentally included commit 186c96f
2. Improved matching of HTTP response codes in `client.GetExpectedOvsdbSchemaVersion`

fnordahl commented on microovn/ovn/cmd/commands.go:
```diff
@@ -200,18 +200,18 @@ func ControllerCtl(s *state.State, args ...string) (string, error) {
+	cert, key := paths.PkiClientCertFiles()
+	arguments := []string{
+		"-t", strconv.Itoa(resultTimeout),
+		"-c", cert,
```
Is this really needed?

The `ovsdb-client` tool is capable of communicating with the cluster through the Unix socket on one of the nodes with a DB member in the cluster.

mkalcok replied:
This addition makes this wrapper usable also on nodes that don't run central services.

fnordahl replied:
That is not a good reason. The test code already defines lists with central nodes and chassis nodes, so you know where you can get the information. The test also defines a "probe container" which also has this information locally.

fnordahl reviewed (changes requested):
Great work! I have now tested an upgrade as well and this process does indeed avoid situations where the end user upgrades nodes in the wrong order. The UX with overview of the entire cluster from `microovn status` is also wonderful.

I have some initial comments and stopped the review at a question on why you need the `ovdb-client` to speak with the SSL sockets, the upstream cluster upgrade code does all communication through a local unix socket on one of the DB nodes.

Thinking about the tests, if we perhaps put some of the code into helper functions, we might fold it into the regular upgrade test file, conditionally enabling the schema tests based on environment, similar to the IPv6 tests for example?

mkalcok commented:
Thanks for the review @fnordahl 

> I have some initial comments and stopped the review at a question on why you need the ovdb-client to speak with the SSL sockets, the upstream cluster upgrade code does all communication through a local unix socket on one of the DB nodes.

main reason is to "make stuff work" on any cluster member, not just those that are running "central" service. Whether it's in the tests or in the `microovn status`, detection of currently running schema would not work on "switch-only" nodes if it tried to use local unix socket.

> Thinking about the tests, if we perhaps put some of the code into helper functions, we might fold it into the regular upgrade test file, conditionally enabling the schema tests based on environment, similar to the IPv6 tests for example?

I'll look into this again, but I'm afraid that it would greatly increase the test complexity. Regular upgrade test is `deploy -> upgrade all nodes -> test if stuff still works`. With these tests in there, we add bunch of other conditions and tests after each step of upgraded cluster member. I'm afraid it could make untangling test failures more difficult. But you do have a point, perhaps there's a way to save some time in this test.

mkalcok commented on tests/test_helper/bats/ovsdb_schema_upgrade.bats:
```diff
@@ -0,0 +1,150 @@
+    # After all cluster members are upgraded, verify that the cluster reports upgraded schema version
+    # as well.
+    local timeout=30
+    local err=1
```
👍  I guess I wasn't sure about exact rules vor variable scoping in bash so I went with safer approach :D

fnordahl commented:
> Thanks for the review @fnordahl
> 
> > I have some initial comments and stopped the review at a question on why you need the ovdb-client to speak with the SSL sockets, the upstream cluster upgrade code does all communication through a local unix socket on one of the DB nodes.
> 
> main reason is to "make stuff work" on any cluster member, not just those that are running "central" service. Whether it's in the tests or in the `microovn status`, detection of currently running schema would not work on "switch-only" nodes if it tried to use local unix socket.

You still have the "just works" capability by using the REST API to call to one of the nodes with the databases?

> > Thinking about the tests, if we perhaps put some of the code into helper functions, we might fold it into the regular upgrade test file, conditionally enabling the schema tests based on environment, similar to the IPv6 tests for example?
> 
> I'll look into this again, but I'm afraid that it would greatly increase the test complexity. Regular upgrade test is `deploy -> upgrade all nodes -> test if stuff still works`. With these tests in there, we add bunch of other conditions and tests after each step of upgraded cluster member. I'm afraid it could make untangling test failures more difficult. But you do have a point, perhaps there's a way to save some time in this test.

There are indeed good arguments to keep them separate too, just wanted to explore the possibilities of code reuse by putting blocks of code that belong together into functions instead of in the body of the test definition.

The foundation of the two tests are very similar. We are in the process of adding a third upgrade test which needs to use the same foundation and just add on data path validation.

fnordahl commented on tests/test_helper/bats/ovsdb_schema_upgrade.bats:
```diff
@@ -0,0 +1,150 @@
+        old_member_msg_sb="$original_sb"
+    fi
+
+    for container in $TEST_CONTAINERS; do
```
This loop iterates over all the containers and moves in a forward direction. This means that the current element is to be upgraded, any previous elements are already upgraded and any coming elements are not upgraded yet. As such we can infer all this information from the current array index.

Would it make sense to use that instead of keeping track of all these extra arrays? Something like:
```bash
read -r -a containers <<< "$TEST_CONTAINERS"
for ((i = 0; i < "${#containers[@]}"; i++)) {
    container=${containers[i]}
    echo "processing container $container"

    for ((j = 0; j < i; j++)) {
        echo "upgraded container ${containers[j]}"
    }

    for ((j = i + 1; j < "${#containers[@]}"; j++)) {
        echo "old container ${containers[j]}"
    }
    echo
}
```

mkalcok replied:
This looks reasonable +1

fnordahl reviewed (changes requested):
To move forward I'd like to move the responsibility of retrieving information about the cluster from the server side API implementation to the client implementation.

In addition to avoiding cascading/scaling issues on the server side it would enable the client to direct queries about the running schema version to the nodes that has this information, removing the need for using credentials and calling over the network with `ovsdb-client`.

On the use of credentials, while your proposal works, we need to consider the consequences of using them in new places. What stands out as problematic to me is the expectation that chassis nodes are allowed privileged access to the NB DB.

The NB API is privileged, and it is intended that only the CMS has access to it. The data path components are expected to only speak to the SB DB.

While there are no restrictions imposed on this today, we need to be able to support stricter requirements around this in the future.

mkalcok commented on microovn/api/ovsdb/schema_version.go:
```diff
@@ -1,64 +1,174 @@
+// cluster member and the version that's actually used. If some members of the cluster expect lower schema version
+// than their peers, it might be a sign that they are running older version of MicroOVN (and by extension, OVN) and
+// need to be upgraded.
+func getSchemaVersionStatus(s *state.State, r *http.Request) response.Response {
```
Core of this implementation can definitely be moved to the client side. Though I'm not entirely clear on what we'd be achieving in terms of scaling. Current implementation results in 1 call from client to server and then `($cluster_size - 1)` calls from the server to other servers in the deployment.
Moving the implementation to client would result in `$cluster_size` number of calls from client to servers. So at the end, there'd be the same number of calls.

mkalcok commented on microovn/api/types/ovsdb.go:
```diff
@@ -0,0 +1,128 @@
+// If all values match and there was no error, this method returns "false".
+//
+// This method returns error if Nodes is an empty list.
+func (s *OvsdbSchemaClusterStatus) UpgradeRequired(ctx context.Context) (bool, error) {
```
This is a method implemented directly on the API result type `OvsdbSchemaClusterStatus`. I think it makes more sense to implement is as method, rather than a free function somewhere else that takes `OvsdbSchemaClusterStatus` as an argument. I'm open to suggestions on alternative implementation, thought it may not matter that much if we end up re-implementing `getSchemaVersionStatus` API, this type may go away entirely.

mkalcok commented on microovn/api/types/ovsdb.go:
```diff
@@ -0,0 +1,128 @@
+// If the cluster does not require update, a simple "OK" is returned. Otherwise, this method returns string
+// that describes currently running OVSDB schema version, schema versions expected by the individual cluster members,
+// and whether there were any errors when fetching the expected schema from other cluster members.
+func (s *OvsdbSchemaClusterStatus) Format(ctx context.Context, cli *client.Client) string {
```
I was thinking about naming it `String`, but again this type may end up going away.

mkalcok commented:
Thank you for the feedback

>On the use of credentials, while your proposal works, we need to consider the consequences of using them in new places. What stands out as problematic to me is the expectation that chassis nodes are allowed privileged access to the NB DB.
>
>The NB API is privileged, and it is intended that only the CMS has access to it. The data path components are expected to only speak to the SB DB.
>
>While there are no restrictions imposed on this today, we need to be able to support stricter requirements around this in the future.

While I didn't fully understand/agree with the argument about the scaling issue of the current implementation. This other argument is imo good enough on its own, to warrant creating MicroOVN API for "currently running schema version". I'll get on it.

mkalcok commented:
Relevant changes since last review:

* MicroOVN no longer uses ssl endpoints when calling `ovsdb-client` wrapper
  * This was needed to get active Southbound/Northbound schema versions, but it was replaced with MicroOVN API endpoint that now provides this value
* API endpoint `getOvsdbClusterStatus` was replaced with `getAllExpectedSchemaVersions` which now focuses only on getting the expected schema version from deployment nodes. It no longer includes active schema version in its reply.
  * Similarly, its API return type was simplified and now only returns list of nodes and their expected schema versions
* Logic around interpreting API return types was moved from `types` package and now lives in the `cmd.microovn` package

One thing that you raised before @fnordahl, was that when request needs to be made to all nodes, client [sends request only to one server](https://github.com/mkalcok/microovn/blob/2f55637f7a5df008036a1d4bbd729473802fa465/microovn/client/client.go#L98), and that server then [fans out requests to the rest of the nodes, collects results, and returns them to original client](https://github.com/mkalcok/microovn/blob/2f55637f7a5df008036a1d4bbd729473802fa465/microovn/api/ovsdb/schema_version.go#L101-L112). That approach was kept in this iteration as well because the MicroOVN's client/command packages (`microovn.client`/`microovn.cmd.microovn`)  are very lightweight. They don't use objects like `state.State` which allows, for example, [to make direct DB queries](https://github.com/mkalcok/microovn/blob/2f55637f7a5df008036a1d4bbd729473802fa465/microovn/node/services.go#L71), they rely purely on communicating via the http API.
This prevents them from easily fetching info about nodes in the cluster and their services. They also can't use very convenient [`State.Cluster()` function that returns client instances for every node in the cluster](https://github.com/mkalcok/microovn/blob/2f55637f7a5df008036a1d4bbd729473802fa465/microovn/api/ovsdb/schema_version.go#L94).
If we really wanted, we could implement APIs for every information about cluster we need. Then in `client`/`cmd` package we could query that API, fetch required information and build client instances manually, but I think that's just unnecessary work.

I also think that it's overall cleaner design to keep code that understands internals on the server side, rather than to have that logic on the client side.

fnordahl commented on microovn/ovn/ovsdb/ovsdb.go:
```diff
@@ -26,3 +41,241 @@ func ExpectedOvsdbSchemaVersion(s *state.State, dbSpec *ovnCmd.OvsdbSpec) (strin
+		return nil
+	}
+
+	backOffMs := 100
```
nit: The number used for back off could be a constant, which would allow the other checks used for time to be multipliers of the constant. If we need to change it, we can change it in one place and the rest of the algorithm will adjust in relation to the constant.

fnordahl commented on microovn/ovn/ovsdb/ovsdb.go:
```diff
@@ -26,3 +41,241 @@ func ExpectedOvsdbSchemaVersion(s *state.State, dbSpec *ovnCmd.OvsdbSpec) (strin
+		}
+
+		time.Sleep(time.Duration(backOffMs) * time.Millisecond)
+		if backOffMs < 2000 {
```
nit: see comment above about using constant and multipliers instead of bare numbers for this algorithm.

fnordahl commented on microovn/ovn/ovsdb/ovsdb.go:
```diff
@@ -26,3 +41,241 @@ func ExpectedOvsdbSchemaVersion(s *state.State, dbSpec *ovnCmd.OvsdbSpec) (strin
+			backOffMs *= 2
+		}
+
+		if backOffMs > 2000 {
+			backOffMs = 2000
```
nit: see comment above about using constant and multipliers instead of bare numbers for this algorithm.

fnordahl reviewed (commented):
Thanks for updating the series!

Not introducing a new place for using credentials and keeping the cluster schema operations local makes it more palatable to me.

On keeping the fan-out behavior in the daemon, it's not my preferred approach and we might get into trouble if there are communication issues to some nodes and/or if users do something silly on large scale clusters.

But to be fair, the underlying plumbing does make use of goroutines for handling requests so we at least won't block, and the API endpoints are relatively light weight (keeping the DB cluster communications local helped there too).

If you deem the cost of moving this to the client to be too high what we have now is an acceptable compromise for me.

mkalcok commented:
I went to to bottom addressing your suggestions @fnordahl and I believe I got them all.

fnordahl commented on tests/test_helper/ovsdb.bash:
```diff
@@ -0,0 +1,81 @@
+            return 1
+    fi
+
+    local pki_dir="/var/snap/microovn/common/data/pki/"
+    local cert="$pki_dir/client-cert.pem"
+    local key="$pki_dir/client-privkey.pem"
+    local ca_cert="$pki_dir/cacert.pem"
```
nit: these are no longer needed

fnordahl reviewed (approved):
Thank you for making the upgrade experience of MicroOVN great!
